README.md
pdfbooktool
===========

A python 3 tool for use in creating your own books.

The input PDF can be of any size, and the default output is an A4 PDF,
ready to print twosided, cut, and bind to A6 books using the "perfect
binding" method. Optionally, one can print on A5 paper, fold, and bind
to an A6 book, using the same method.

This package installs a script called 'booktool' and this is what you would 
actually use for preparing your PDFs.


Usage
-----

In a terminal, navigate to the directory of your PDF, then do:

```
$ booktool input-file.pdf
```

Or optionally:

```
$ booktool --a5paper input-file.pdf
```

For A5-sized paper output.

This will output a couple of intermediary files (which will be removed once the
script is finished running) into the working directory, and eventually a file
called "out.pdf". This new file is a reordered PDF, ready for double-sided 
printing. Use "long edge binding" as your double-sided setting on 
your printer for A4. If you opted for A5 paper, use "short edge" instead.

Once printed, you can cut the sheets in half (only needed if you print in A4) 
and bind them together using a technique called "perfect binding". There exist
several videos on youtube that explains how to do this.

A good open-source tool for creating suitable PDFs is 'scribus'.


Requirements
------------

This package requires python 3, and the library 'PyPDF2' to work, the install 
process should install this automatically if needed.


Installation
------------

The most convenient way to install pdfbooktool is with pip. Since the script 
currently doesn't seem to work on Windows, installation instructions are only 
given for linux:

Make sure you have python 3 installed, then open a terminal and type:

```
$ sudo pip3 install pdfbooktool
```

This should automatically install the tool, and it's dependencies.


Running under windows
---------------------

Running this script under Windows has barely been testet. Neither PyPDF2 nor
pdfbooktool should use any features that are platform specific, however it 
doesn't seem to currently work under Windows 7. Help with further testing and
making it work, is wanted.


Changelog.txt
v0.1.5
    - Added support for outputing A5 PDFs for A6 perfect binding.
v0.1.4
    - Added dynamic scaling of input PDFs. Now the input PDF can be any size;
      however, for best results it should be A6.
v0.1.3
    - Fixed a bug that caused the two lower A6 pages on each A4 page to be
      switched.
v0.1.2
    - Fixed an issue with the package structure. The script did not work once
      installed before this.
v0.1.1
    - Changed how the usage-message from the script is displayed.
v0.1.0
    - First release to pypi and GitHub


pdfbooktool/pdfbooktool.py
#!/usr/bin/python3

# pdfbooktool.py - functions for preparing PDFs for book printing.
#    
# Copyright (C) 2016  Svein-Kåre Bjørnsen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# TOOLBOX
##################################
from math import ceil

def ceil_iter(iterable):
    """Takes iterable of floats and returns list of ceil'ed ints"""
    return [int(ceil(elem)) for elem in iterable]

##################################

from PyPDF2 import PdfFileWriter, PdfFileReader

# Reordering schemes
SCHEME_A6_PERFECT = 0
SCHEME_A5_PERFECT = 1

# PDF-specific size dimensions
UNIT_SIZE_A4 = (595.28000, 841.89000)
UNIT_SIZE_A6 = (297.64063, 419.53039)

class PageError(Exception):
    pass

def shuffle_pages(num, scheme):
    """ Takes a number (pages) and the reordering scheme to use.
    Returns a list of indexes to use with page reordering. """
    
    book = []
    counter = 0

    if scheme == SCHEME_A6_PERFECT:
        sequence = [3, -1, 5, 1, -3, -3, -1, -1]
    elif scheme == SCHEME_A5_PERFECT:
        sequence = [3, -1, -1, -1]

    for page in range(num):
        book.append(0)
        
    for page in range(num):
        if counter == len(sequence):
            counter = 0
        book[page] = page + sequence[counter]
        counter += 1

    return book

def calc_pages_to_add(pdfreader, scheme):
    """ Takes a PdfFileReader and scheme. Returns number of pages
        to add for the selected scheme. """
    
    if scheme == SCHEME_A6_PERFECT:
        counter = 0
        pages = pdfreader.getNumPages()
        done = False
        while not done:
            if pages % 8 != 0:
                counter += 1
                pages += 1
            else:
                done = True
        return counter
    else:
        return None

    # TODO A5_PERFECT

def add_blank_pages(pdf_file):
    """ Takes filename of pdf and number of blank pages to append.
        Writes to "output.pdf" in same directory."""

    writer = PdfFileWriter()
    pdf = PdfFileReader(open(pdf_file, "rb"))
    
    for page in pdf.pages:
        writer.addPage(page)

    for page in range(calc_pages_to_add(pdf, SCHEME_A6_PERFECT)):
        writer.addBlankPage()

    with open("aux.pdf", "wb") as f:
        writer.write(f)

def reorder_pages(pdfreader, scheme):
    """ Takes PdfFileReader and reordering scheme. Returns 
        PdfFileWriter, ordered according to scheme. """
    
    pageorder = shuffle_pages( pdfreader.getNumPages(),
                               scheme )
    out = PdfFileWriter()

    for page in pageorder:
        out.addPage(pdfreader.pages[page])
    
    return out

def add_blank_page(writer, unit_size):
    page = writer.addBlankPage(unit_size[0], unit_size[1])
    return page
    
def make_page(writer, reader, offset=0):
    tx = UNIT_SIZE_A6[0]
    ty = UNIT_SIZE_A6[1]
    
    page = add_blank_page(writer, UNIT_SIZE_A4)
    
    page.mergeTranslatedPage(reader.pages[0+offset], 0, ty)
    page.mergeTranslatedPage(reader.pages[1+offset], tx, ty)
    page.mergeTranslatedPage(reader.pages[2+offset], tx, 0)
    page.mergeTranslatedPage(reader.pages[3+offset], 0, 0)
    
def a6_to_a4_merge(pdf_file):
    
    pdf = PdfFileReader(open(pdf_file, "rb"))
    
    if ceil_iter(pdf.pages[0].mediaBox.upperRight) != ceil_iter(UNIT_SIZE_A6):
        raise PageError("Trying to merge PDF that is not A6 as A6")
        
    writer = PdfFileWriter()
    
    number_pages = pdf.getNumPages()
    if number_pages % 4 != 0:
        raise PageError("Number of pages in PDF not divisible by 4")
    
    page_offset = 0
    while page_offset < number_pages:
        make_page(writer, pdf, page_offset)
        page_offset += 4
    
    with open("out.pdf", "wb") as f:
        writer.write(f)



pdfbooktool/functions.py
#!/usr/bin/python3

# pdfbooktool.py - functions for preparing PDFs for book printing.
#    
# Copyright (C) 2016  Svein-Kåre Bjørnsen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# TOOLBOX
##################################
from math import ceil

def ceil_iter(iterable):
    """Takes iterable of floats and returns list of ceil'ed ints"""
    return [int(ceil(elem)) for elem in iterable]

##################################

from PyPDF2 import PdfFileWriter, PdfFileReader

# Reordering schemes
SCHEME_A6_PERFECT = 0
SCHEME_A5_PERFECT = 1  # Works for both A4 paper -> A5 perfect and
                       # A5 paper -> A6 perfect

# PDF-specific size dimensions
UNIT_SIZE_A4 = (595.28000, 841.89000)
UNIT_SIZE_A5 = (419.53039, 595.28000)
UNIT_SIZE_A6 = (297.64063, 419.53039)

class PageError(Exception):
    pass

def shuffle_pages(num, scheme):
    """ Takes a number (pages) and the reordering scheme to use.
    Returns a list of indexes to use with page reordering. """
    
    book = []
    counter = 0

    if scheme == SCHEME_A6_PERFECT:
        sequence = [3, -1, 5, 1, -3, -3, -1, -1]
    elif scheme == SCHEME_A5_PERFECT:
        sequence = [3, -1, -1, -1]
    
    for page in range(num):
        book.append(0)
        
    for page in range(num):
        if counter == len(sequence):
            counter = 0
        book[page] = page + sequence[counter]
        counter += 1

    return book

def calc_pages_to_add(pdfreader, scheme):
    """ Takes a PdfFileReader and scheme. Returns number of pages
        to add for the selected scheme. """
    
    if scheme == SCHEME_A6_PERFECT:
        counter = 0
        pages = pdfreader.getNumPages()
        done = False
        while not done:
            if pages % 8 != 0:
                counter += 1
                pages += 1
            else:
                done = True
        return counter
    elif scheme is SCHEME_A5_PERFECT:
        counter = 0
        pages = pdfreader.getNumPages()
        done = False
        while not done:
            if pages % 4 != 0:
                counter += 1
                pages += 1
            else:
                done = True
        return counter
    else:
        return None


def add_blank_pages(pdf_file, scheme):
    """ Takes filename of pdf and number of blank pages to append.
        Writes to "output.pdf" in same directory."""

    writer = PdfFileWriter()
    pdf = PdfFileReader(open(pdf_file, "rb"))
    
    for page in pdf.pages:
        writer.addPage(page)

    for page in range(calc_pages_to_add(pdf, scheme)):
        writer.addBlankPage()

    with open("aux.pdf", "wb") as f:
        writer.write(f)

def reorder_pages(pdfreader, scheme):
    """ Takes PdfFileReader and reordering scheme. Returns 
        PdfFileWriter, ordered according to scheme. """
    
    pageorder = shuffle_pages( pdfreader.getNumPages(),
                               scheme )
    out = PdfFileWriter()

    for page in pageorder:
        out.addPage(pdfreader.pages[page])
    
    return out

def add_blank_page(writer, unit_size):
    page = writer.addBlankPage(unit_size[0], unit_size[1])
    return page
    
def make_page(writer, reader, unit_size_in, unit_size_out, offset=0):

    if unit_size_out is UNIT_SIZE_A4:
        page = add_blank_page(writer, unit_size_out)
        tx = unit_size_in[0]
        ty = unit_size_in[1]
        page.mergeTranslatedPage(reader.pages[0+offset], 0, ty)
        page.mergeTranslatedPage(reader.pages[1+offset], tx, ty)
        page.mergeTranslatedPage(reader.pages[2+offset], 0, 0)
        page.mergeTranslatedPage(reader.pages[3+offset], tx, 0)
    elif unit_size_out is UNIT_SIZE_A5:
        tx = unit_size_in[0]
        ty = unit_size_in[1]
        page = add_blank_page(writer, (UNIT_SIZE_A5[1], UNIT_SIZE_A5[0]))
        pageone = reader.pages[0+offset]
        pagetwo = reader.pages[1+offset]

        page.mergeTranslatedPage(pageone,0,0)
        page.mergeTranslatedPage(pagetwo,tx,0)

        page.rotateClockwise(90)

    
def scale_to_size(reader, unit_size):
    # Create a writer and populate with pages at unit_size.
    # Scale and merge pages from reader, centered, onto these pages.
    # return open BytesIO object written to by writer
    
    # For now, just write to aux.pdf 
    
    # determine scaling factor:
    input_unit_size = reader.pages[0].mediaBox.upperRight
    
    factors = [a / b for a, b in zip(unit_size, input_unit_size)]
    if factors[0] <= factors[1]:
        # Use x-factor for scaling if this is lowest, or equal to, y-factor
        factor = factors[0]
    else:
        # Otherwise, use the y-factor
        factor = factors[1]

    writer = PdfFileWriter()
    
    for in_page in reader.pages:
        out_page = add_blank_page(writer, unit_size)
        out_page.mergeScaledPage(in_page, factor, expand=True)
    
    with open("aux.pdf", "wb") as f:    
        writer.write(f)
    
def a6_to_a4_merge(pdf_file):
    
    pdf = PdfFileReader(open(pdf_file, "rb"))

    # Check if pdf size is A6, if not, scale.
    if ceil_iter(pdf.pages[0].mediaBox.upperRight) != ceil_iter(UNIT_SIZE_A6):
        scale_to_size(pdf, UNIT_SIZE_A6)
        pdf = PdfFileReader(open("aux.pdf", "rb"))
        
    writer = PdfFileWriter()
    
    number_pages = pdf.getNumPages()
    if number_pages % 4 != 0:
        raise PageError("Number of pages in PDF not divisible by 4")
    
    page_offset = 0
    while page_offset < number_pages:
        make_page(writer, pdf, UNIT_SIZE_A6, UNIT_SIZE_A4, page_offset)
        page_offset += 4
    
    with open("out.pdf", "wb") as f:
        writer.write(f)

def a6_to_a5_merge(pdf_file):

    pdf = PdfFileReader(open(pdf_file, "rb"))

    # Check if pdf size is A6, if not, scale.
    if ceil_iter(pdf.pages[0].mediaBox.upperRight) != ceil_iter(UNIT_SIZE_A6):
        scale_to_size(pdf, UNIT_SIZE_A6)
        pdf = PdfFileReader(open("aux.pdf", "rb"))
        
    writer = PdfFileWriter()
    
    number_pages = pdf.getNumPages()
    if number_pages % 2 != 0:
        raise PageError("Number of pages in PDF not divisible by 2")
    
    page_offset = 0
    while page_offset < number_pages:
        make_page(
            writer,
            pdf,
            UNIT_SIZE_A6,
            UNIT_SIZE_A5,
            offset=page_offset
        )
        page_offset += 2 
    
    with open("out.pdf", "wb") as f:
        writer.write(f)
